File: apps/strategy-service/src/observability/metrics.py
"""Prometheus Metrics for Strategy Service."""

import os
from prometheus_client import Counter, Histogram, Gauge, CollectorRegistry, generate_latest
from aiohttp import web

# Global registry
REGISTRY = CollectorRegistry()

# Strategy metrics
strategy_signals = Counter(
    "enthropic_strategy_signals_total",
    "Total strategy signals generated",
    ["strategy", "side", "symbol"],
    registry=REGISTRY,
)

strategy_execution_duration = Histogram(
    "enthropic_strategy_execution_duration_seconds",
    "Strategy execution duration",
    ["strategy"],
    buckets=[0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0],
    registry=REGISTRY,
)

active_strategies = Gauge(
    "enthropic_active_strategies",
    "Number of active strategies",
    registry=REGISTRY,
)

# Resilience metrics
circuit_breaker_state = Gauge(
    "enthropic_circuit_breaker_state",
    "Circuit breaker state",
    ["name"],
    registry=REGISTRY,
)

retry_attempts = Counter(
    "enthropic_retry_attempts_total",
    "Retry attempts",
    ["operation", "outcome"],
    registry=REGISTRY,
)


class TradingMetrics:
    """Metrics accessor class."""
    
    strategy_signals = strategy_signals
    strategy_execution_duration = strategy_execution_duration
    active_strategies = active_strategies
    circuit_breaker_state = circuit_breaker_state
    retry_attempts = retry_attempts


_metrics = TradingMetrics()


def init_metrics() -> TradingMetrics:
    """Initialize metrics."""
    return _metrics


def get_metrics() -> TradingMetrics:
    """Get metrics instance."""
    return _metrics


class MetricsServer:
    """HTTP server for metrics and health endpoints."""
    
    def __init__(self, port: int = 9102):
        self.port = port
        self.app = web.Application()
        self.app.router.add_get("/metrics", self.metrics_handler)
        self.app.router.add_get("/health", self.health_handler)
        self._healthy = True
    
    async def metrics_handler(self, request: web.Request) -> web.Response:
        return web.Response(
            body=generate_latest(REGISTRY),
            content_type="text/plain",
        )
    
    async def health_handler(self, request: web.Request) -> web.Response:
        if self._healthy:
            return web.json_response({"status": "healthy"})
        return web.json_response({"status": "unhealthy"}, status=503)
    
    def set_healthy(self, healthy: bool) -> None:
        self._healthy = healthy
    
    async def start(self) -> None:
        runner = web.AppRunner(self.app)
        await runner.setup()
        site = web.TCPSite(runner, "0.0.0.0", self.port)
        await site.start()

File: apps/execution-core/src/observability/metrics.rs
//! Prometheus Metrics for Trading Platform
//! Custom metrics for order processing, positions, and system health

use once_cell::sync::Lazy;
use prometheus::{
    Counter, CounterVec, Gauge, GaugeVec, HistogramVec,
    Opts, Registry, TextEncoder, Encoder,
};
use std::sync::Mutex;

/// Global metrics registry
static REGISTRY: Lazy<Registry> = Lazy::new(Registry::new);

/// Application metrics
pub struct Metrics {
    pub orders_processed_total: CounterVec,
    pub orders_rejected_total: CounterVec,
    pub order_processing_duration: HistogramVec,
    pub position_updates_total: Counter,
    pub active_positions: Gauge,
    pub position_pnl: GaugeVec,
    pub db_pool_connections: GaugeVec,
    pub nats_messages_received: CounterVec,
    pub nats_messages_published: CounterVec,
    pub circuit_breaker_state: GaugeVec,
    pub retry_attempts_total: CounterVec,
}

static METRICS: Lazy<Mutex<Option<Metrics>>> = Lazy::new(|| Mutex::new(None));

/// Initialize metrics
pub fn init_metrics(service_name: &str) -> anyhow::Result<()> {
    let orders_processed_total = CounterVec::new(
        Opts::new("enthropic_orders_processed_total", "Total orders processed")
            .namespace("enthropic")
            .const_label("service", service_name),
        &["status", "side", "symbol"]
    )?;

    let orders_rejected_total = CounterVec::new(
        Opts::new("enthropic_orders_rejected_total", "Total orders rejected")
            .namespace("enthropic")
            .const_label("service", service_name),
        &["reason"]
    )?;

    let order_processing_duration = HistogramVec::new(
        prometheus::HistogramOpts::new(
            "enthropic_order_processing_duration_seconds",
            "Order processing latency in seconds"
        )
            .namespace("enthropic")
            .const_label("service", service_name)
            .buckets(vec![0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0]),
        &["operation"]
    )?;

    let position_updates_total = Counter::new(
        "enthropic_position_updates_total",
        "Total position updates"
    )?;

    let active_positions = Gauge::new(
        "enthropic_active_positions",
        "Number of active positions"
    )?;

    let position_pnl = GaugeVec::new(
        Opts::new("enthropic_position_pnl", "Position PnL by type"),
        &["type"] // realized, unrealized
    )?;

    let db_pool_connections = GaugeVec::new(
        Opts::new("enthropic_db_pool_connections", "Database pool connections"),
        &["state"] // active, idle
    )?;

    let nats_messages_received = CounterVec::new(
        Opts::new("enthropic_nats_messages_received_total", "NATS messages received"),
        &["subject"]
    )?;

    let nats_messages_published = CounterVec::new(
        Opts::new("enthropic_nats_messages_published_total", "NATS messages published"),
        &["subject"]
    )?;

    let circuit_breaker_state = GaugeVec::new(
        Opts::new("enthropic_circuit_breaker_state", "Circuit breaker state (0=closed, 0.5=half-open, 1=open)"),
        &["name"]
    )?;

    let retry_attempts_total = CounterVec::new(
        Opts::new("enthropic_retry_attempts_total", "Total retry attempts"),
        &["operation", "outcome"]
    )?;

    // Register all metrics
    REGISTRY.register(Box::new(orders_processed_total.clone()))?;
    REGISTRY.register(Box::new(orders_rejected_total.clone()))?;
    REGISTRY.register(Box::new(order_processing_duration.clone()))?;
    REGISTRY.register(Box::new(position_updates_total.clone()))?;
    REGISTRY.register(Box::new(active_positions.clone()))?;
    REGISTRY.register(Box::new(position_pnl.clone()))?;
    REGISTRY.register(Box::new(db_pool_connections.clone()))?;
    REGISTRY.register(Box::new(nats_messages_received.clone()))?;
    REGISTRY.register(Box::new(nats_messages_published.clone()))?;
    REGISTRY.register(Box::new(circuit_breaker_state.clone()))?;
    REGISTRY.register(Box::new(retry_attempts_total.clone()))?;

    let metrics = Metrics {
        orders_processed_total,
        orders_rejected_total,
        order_processing_duration,
        position_updates_total,
        active_positions,
        position_pnl,
        db_pool_connections,
        nats_messages_received,
        nats_messages_published,
        circuit_breaker_state,
        retry_attempts_total,
    };

    let mut guard = METRICS.lock().unwrap();
    *guard = Some(metrics);

    tracing::info!("Prometheus metrics initialized");
    Ok(())
}

/// Get metrics instance
pub fn get_metrics() -> std::sync::MutexGuard<'static, Option<Metrics>> {
    METRICS.lock().unwrap()
}

/// Encode metrics to Prometheus text format
pub fn encode_metrics() -> String {
    let encoder = TextEncoder::new();
    let metric_families = REGISTRY.gather();
    let mut buffer = Vec::new();
    encoder.encode(&metric_families, &mut buffer).unwrap_or_default();
    String::from_utf8(buffer).unwrap_or_default()
}